src/sql/src/plan/statement/ddl.rs

```diff
@@ -3697,6 +3690,182 @@ pub fn describe_alter_index_options(
+                            "cannot drop {} without CASCADE: still depended upon by non-owned catalog items {}",
+                            scx.catalog.resolve_full_name(item.name()).to_string().quoted(),
+                            names.join(", ")
+                        );
+                    }
+                }
+            }
+            drop_ids.push(item.id().into());
+        }
+        update_revokes(item.id().into(), item.privileges(), &role_ids, &mut revokes);
+    }
+
+    // Schemas
+    for schema in scx.catalog.get_schemas() {
+        if !schema.id().is_temporary() {
+            if role_ids.contains(&schema.owner_id()) {
+                if !cascade {
+                    let non_owned_dependencies: Vec<_> = schema
+                        .item_ids()
+                        .values()
+                        .map(|global_id| scx.catalog.get_item(global_id))
+                        .filter(|item| dependency_prevents_drop(item.item_type().into(), *item))
+                        .filter(|item| !role_ids.contains(&item.owner_id()))
+                        .collect();
+                    if !non_owned_dependencies.is_empty() {
+                        let full_schema_name = scx.catalog.resolve_full_schema_name(schema.name());
+                        sql_bail!(
+                            "schema {} cannot be dropped without CASCADE while it contains non-owned objects",
+                            full_schema_name.to_string().quoted()
+                        );
+                    }
+                }
+                drop_ids.push((*schema.database(), *schema.id()).into())
+            }
+            update_revokes(
+                (*schema.database(), *schema.id()).into(),
+                schema.privileges(),
+                &role_ids,
+                &mut revokes,
+            );
+        }
+    }
+
+    // Databases
+    for database in scx.catalog.get_databases() {
+        if role_ids.contains(&database.owner_id()) {
+            if !cascade {
+                let non_owned_schemas: Vec<_> = database
+                    .schemas()
+                    .into_iter()
+                    .filter(|schema| !role_ids.contains(&schema.owner_id()))
+                    .collect();
+                if !non_owned_schemas.is_empty() {
+                    sql_bail!(
+                        "database {} cannot be dropped without CASCADE while it contains non-owned schemas",
+                        database.name().quoted(),
+                    );
+                }
+            }
+            drop_ids.push(database.id().into());
+        }
+        update_revokes(
+            database.id().into(),
+            database.privileges(),
+            &role_ids,
+            &mut revokes,
+        );
+    }
+
+    let drop_ids = scx.catalog.object_dependents(&drop_ids);
+
+    let system_ids: Vec<_> = drop_ids.iter().filter(|id| id.is_system()).collect();
+    if !system_ids.is_empty() {
```

Comment:
Could test this block:
```
mz_system=> drop owned by mz_system;
ERROR:  cannot drop r1, r1, mz_system, mz_introspection, pg_catalog.pg_attribute, pg_catalog.pg_type, mz_catalog.mz_array_types, mz_internal.mz_cluster_replica_history, mz_catalog.mz_audit_events, mz_catalog.mz_aws_privatelink_connections, mz_catalog.mz_base_types, mz_internal.mz_cluster_replica_utilization, mz_internal.mz_show_cluster_replicas_ind, mz_internal.mz_show_cluster_replicas, mz_internal.mz_cluster_replicas_ind, mz_catalog.mz_cluster_replicas, mz_internal.mz_show_materialized_views_ind, mz_internal.mz_show_materialized_views, mz_internal.mz_show_indexes_ind, mz_internal.mz_show_indexes, mz_internal.mz_show_clusters_ind, mz_internal.mz_clusters_ind, mz_catalog.mz_clusters, pg_catalog.pg_attrdef, information_schema.columns, mz_internal.mz_show_columns_ind, mz_catalog.mz_columns, mz_internal.mz_dataflow_arrangement_sizes, mz_internal.mz_show_all_objects_ind, mz_catalog.mz_objects, mz_internal.mz_show_connections_ind, mz_catalog.mz_connections, pg_catalog.pg_tables, pg_catalog.pg_namespace, pg_catalog.pg_description, pg_catalog.pg_class, pg_catalog.pg_database, pg_catalog.pg_index, pg_catalog.pg_proc, pg_catalog.pg_views, pg_catalog.pg_matviews, information_schema.tables, mz_internal.mz_show_databases_ind, mz_catalog.mz_databases, mz_catalog.mz_egress_ips, mz_catalog.mz_functions, mz_catalog.mz_index_columns, mz_catalog.mz_indexes, mz_catalog.mz_kafka_connections, mz_catalog.mz_kafka_sinks, mz_catalog.mz_list_types, mz_catalog.mz_map_types, mz_catalog.mz_relations, mz_catalog.mz_materialized_views, pg_catalog.pg_operator, mz_catalog.mz_operators, mz_catalog.mz_pseudo_types, pg_catalog.pg_auth_members, mz_catalog.mz_role_members, pg_catalog.pg_roles, pg_catalog.pg_authid, mz_catalog.mz_roles, mz_internal.mz_show_schemas_ind, mz_catalog.mz_schemas, mz_internal.mz_show_secrets_ind, mz_catalog.mz_secrets, mz_internal.mz_sink_statuses, mz_internal.mz_show_sinks_ind, mz_catalog.mz_sinks, mz_internal.mz_source_statuses_ind, mz_internal.mz_source_statuses, mz_internal.mz_show_sources_ind, mz_catalog.mz_sources, mz_catalog.mz_ssh_tunnel_connections, mz_catalog.mz_storage_usage, mz_internal.mz_show_tables_ind, mz_catalog.mz_tables, mz_internal.mz_show_types_ind, mz_catalog.mz_types, mz_internal.mz_show_views_ind, mz_catalog.mz_views, pg_catalog._bool, pg_catalog._bpchar, pg_catalog._bytea, pg_catalog._char, pg_catalog._date, pg_catalog._daterange, pg_catalog._float4, pg_catalog._float8, pg_catalog._int2, pg_catalog._int2vector, pg_catalog._int4, pg_catalog._int4range, pg_catalog._int8, pg_catalog._int8range, pg_catalog._interval, pg_catalog._jsonb, pg_catalog._mz_aclitem, pg_catalog._mz_timestamp, pg_catalog._numeric, pg_catalog._numrange, pg_catalog._oid, pg_catalog._record, pg_catalog._regclass, pg_catalog._regproc, pg_catalog._regtype, pg_catalog._text, pg_catalog._time, pg_catalog._timestamp, pg_catalog._timestamptz, pg_catalog._tsrange, pg_catalog._tstzrange, pg_catalog._uint2, pg_catalog._uint4, pg_catalog._uint8, pg_catalog._uuid, pg_catalog._varchar, pg_catalog.any, pg_catalog.anyarray, pg_catalog.anycompatible, pg_catalog.anycompatiblearray, pg_catalog.anycompatiblelist, pg_catalog.anycompatiblemap, pg_catalog.anycompatiblenonarray, pg_catalog.anycompatiblerange, pg_catalog.anyelement, pg_catalog.anynonarray, pg_catalog.anyrange, pg_catalog.pg_constraint, pg_catalog.pg_collation, pg_catalog.pg_policy, pg_catalog.pg_inherits, pg_catalog.bool, pg_catalog.bpchar, pg_catalog.bytea, pg_catalog.pg_am, pg_catalog.char, pg_catalog.date, pg_catalog.daterange, pg_catalog.pg_enum, pg_catalog.float4, pg_catalog.float8, pg_catalog.int2, pg_catalog.int2vector, pg_catalog.int4, pg_catalog.int4range, pg_catalog.int8, pg_catalog.int8range, pg_catalog.interval, pg_catalog.jsonb, pg_catalog.list, pg_catalog.map, pg_catalog.mz_aclitem, pg_catalog.mz_timestamp, pg_catalog.numeric, pg_catalog.numrange, pg_catalog.pg_range, pg_catalog.oid, pg_catalog.pg_settings, pg_catalog.record, pg_catalog.regclass, pg_catalog.regproc, pg_catalog.regtype, pg_catalog.text, pg_catalog.time, pg_catalog.timestamp, pg_catalog.timestamptz, pg_catalog.tsrange, pg_catalog.tstzrange, pg_catalog.uint2, pg_catalog.uint4, pg_catalog.uint8, pg_catalog.uuid, pg_catalog.varchar, mz_internal.mz_active_peeks, mz_internal.mz_active_peeks_per_worker_s1_primary_idx, mz_internal.mz_active_peeks_per_worker_s2_primary_idx, mz_internal.mz_active_peeks_per_worker_u1_primary_idx, mz_internal.mz_active_peeks_per_worker, mz_internal.mz_arrangement_sizes, mz_internal.mz_arrangement_sizes_per_worker, mz_internal.mz_arrangement_batches_raw_s1_primary_idx, mz_internal.mz_arrangement_batches_raw_s2_primary_idx, mz_internal.mz_arrangement_batches_raw_u1_primary_idx, mz_internal.mz_arrangement_batches_raw, mz_internal.mz_records_per_dataflow_operator, mz_internal.mz_records_per_dataflow, mz_internal.mz_records_per_dataflow_per_worker, mz_internal.mz_records_per_dataflow_operator_per_worker, mz_internal.mz_arrangement_records_raw_s1_primary_idx, mz_internal.mz_arrangement_records_raw_s2_primary_idx, mz_internal.mz_arrangement_records_raw_u1_primary_idx, mz_internal.mz_arrangement_records_raw, mz_internal.mz_arrangement_sharing, mz_internal.mz_arrangement_sharing_per_worker, mz_internal.mz_arrangement_sharing_raw_s1_primary_idx, mz_internal.mz_arrangement_sharing_raw_s2_primary_idx, mz_internal.mz_arrangement_sharing_raw_u1_primary_idx, mz_internal.mz_arrangement_sharing_raw, mz_internal.mz_cluster_links, mz_internal.mz_cluster_replica_frontiers, mz_internal.mz_cluster_replica_heartbeats, mz_internal.mz_cluster_replica_metrics_ind, mz_internal.mz_cluster_replica_metrics, mz_internal.mz_cluster_replica_sizes_ind, mz_internal.mz_cluster_replica_sizes, mz_internal.mz_cluster_replica_statuses_ind, mz_internal.mz_cluster_replica_statuses, mz_internal.mz_compute_delays_histogram, mz_internal.mz_compute_delays_histogram_per_worker, mz_internal.mz_compute_delays_histogram_raw_s1_primary_idx, mz_internal.mz_compute_delays_histogram_raw_s2_primary_idx, mz_internal.mz_compute_delays_histogram_raw_u1_primary_idx, mz_internal.mz_compute_delays_histogram_raw, mz_internal.mz_compute_dependencies, mz_internal.mz_compute_dependencies_per_worker_s1_primary_idx, mz_internal.mz_compute_dependencies_per_worker_s2_primary_idx, mz_internal.mz_compute_dependencies_per_worker_u1_primary_idx, mz_internal.mz_compute_dependencies_per_worker, mz_internal.mz_compute_exports, mz_internal.mz_compute_exports_per_worker_s1_primary_idx, mz_internal.mz_compute_exports_per_worker_s2_primary_idx, mz_internal.mz_compute_exports_per_worker_u1_primary_idx, mz_internal.mz_compute_exports_per_worker, mz_internal.mz_compute_frontiers, mz_internal.mz_compute_frontiers_per_worker_s1_primary_idx, mz_internal.mz_compute_frontiers_per_worker_s2_primary_idx, mz_internal.mz_compute_frontiers_per_worker_u1_primary_idx, mz_internal.mz_compute_frontiers_per_worker, mz_internal.mz_compute_import_frontiers, mz_internal.mz_compute_import_frontiers_per_worker_s1_primary_idx, mz_internal.mz_compute_import_frontiers_per_worker_s2_primary_idx, mz_internal.mz_compute_import_frontiers_per_worker_u1_primary_idx, mz_internal.mz_compute_import_frontiers_per_worker, mz_internal.mz_compute_operator_durations_histogram, mz_internal.mz_compute_operator_durations_histogram_per_worker, mz_internal.mz_compute_operator_durations_histogram_raw_s1_primary_idx, mz_internal.mz_compute_operator_durations_histogram_raw_s2_primary_idx, mz_internal.mz_compute_operator_durations_histogram_raw_u1_primary_idx, mz_internal.mz_compute_operator_durations_histogram_raw, mz_internal.mz_dataflow_addresses, mz_internal.mz_dataflows, mz_internal.mz_dataflow_operator_dataflows, mz_internal.mz_dataflow_operator_dataflows_per_worker, mz_internal.mz_dataflows_per_worker, mz_internal.mz_dataflow_operator_parents, mz_internal.mz_dataflow_operator_parents_per_worker, mz_internal.mz_dataflow_channel_operators, mz_internal.mz_dataflow_channel_operators_per_worker, mz_internal.mz_dataflow_addresses_per_worker_s1_primary_idx, mz_internal.mz_dataflow_addresses_per_worker_s2_primary_idx, mz_internal.mz_dataflow_addresses_per_worker_u1_primary_idx, mz_internal.mz_dataflow_addresses_per_worker, mz_internal.mz_dataflow_channels, mz_internal.mz_dataflow_channels_per_worker_s1_primary_idx, mz_internal.mz_dataflow_channels_per_worker_s2_primary_idx, mz_internal.mz_dataflow_channels_per_worker_u1_primary_idx, mz_internal.mz_dataflow_channels_per_worker, mz_internal.mz_dataflow_operator_reachability, mz_internal.mz_dataflow_operator_reachability_per_worker, mz_internal.mz_dataflow_operator_reachability_raw_s1_primary_idx, mz_internal.mz_dataflow_operator_reachability_raw_s2_primary_idx, mz_internal.mz_dataflow_operator_reachability_raw_u1_primary_idx, mz_internal.mz_dataflow_operator_reachability_raw, mz_internal.mz_dataflow_operators, mz_internal.mz_dataflow_operators_per_worker_s1_primary_idx, mz_internal.mz_dataflow_operators_per_worker_s2_primary_idx, mz_internal.mz_dataflow_operators_per_worker_u1_primary_idx, mz_internal.mz_dataflow_operators_per_worker, mz_internal.mz_message_counts, mz_internal.mz_message_counts_per_worker, mz_internal.mz_message_counts_received_raw_s1_primary_idx, mz_internal.mz_message_counts_received_raw_s2_primary_idx, mz_internal.mz_message_counts_received_raw_u1_primary_idx, mz_internal.mz_message_counts_received_raw, mz_internal.mz_message_counts_sent_raw_s1_primary_idx, mz_internal.mz_message_counts_sent_raw_s2_primary_idx, mz_internal.mz_message_counts_sent_raw_u1_primary_idx, mz_internal.mz_message_counts_sent_raw, mz_internal.mz_object_dependencies, mz_internal.mz_peek_durations_histogram, mz_internal.mz_peek_durations_histogram_per_worker, mz_internal.mz_peek_durations_histogram_raw_s1_primary_idx, mz_internal.mz_peek_durations_histogram_raw_s2_primary_idx, mz_internal.mz_peek_durations_histogram_raw_u1_primary_idx, mz_internal.mz_peek_durations_histogram_raw, mz_internal.mz_postgres_sources, mz_internal.mz_scheduling_elapsed, mz_internal.mz_scheduling_elapsed_per_worker, mz_internal.mz_scheduling_elapsed_raw_s1_primary_idx, mz_internal.mz_scheduling_elapsed_raw_s2_primary_idx, mz_internal.mz_scheduling_elapsed_raw_u1_primary_idx, mz_internal.mz_scheduling_elapsed_raw, mz_internal.mz_scheduling_parks_histogram, mz_internal.mz_scheduling_parks_histogram_per_worker, mz_internal.mz_scheduling_parks_histogram_raw_s1_primary_idx, mz_internal.mz_scheduling_parks_histogram_raw_s2_primary_idx, mz_internal.mz_scheduling_parks_histogram_raw_u1_primary_idx, mz_internal.mz_scheduling_parks_histogram_raw, mz_internal.mz_sessions, mz_internal.mz_sink_statistics, mz_internal.mz_sink_status_history, mz_internal.mz_source_statistics, mz_internal.mz_source_status_history_ind, mz_internal.mz_source_status_history, mz_internal.mz_storage_shards, mz_internal.mz_storage_usage_by_shard, mz_internal.mz_subscriptions, mz_internal.mz_view_foreign_keys, mz_internal.mz_view_keys, mz_catalog, pg_catalog, mz_internal, information_schema because they are required by the database system
```
Some of the names seem a bit confusing, `r1` is the replica name I guess, exists twice, might want to add what cluster it belongs to?